File: ios/GamezopPostMessage/EventStore.swift
import Foundation

@MainActor
final class EventStore: ObservableObject {
    @Published private(set) var events: [CapturedGameEvent] = []
    @Published var selectedEventID: Int64?
    @Published private(set) var droppedEventCount = 0

    private var nextID: Int64 = 1
    private let capacity: Int

    init(capacity: Int = 500) {
        self.capacity = capacity
    }

    func capture(_ body: Any, at date: Date = Date()) {
        let event = GameEventParser.parse(body: body, id: nextID, receivedAt: date)
        nextID += 1
        events.append(event)
        if events.count > capacity {
            let overflow = events.count - capacity
            events.removeFirst(overflow)
            droppedEventCount += overflow
        }
        selectedEventID = event.id
    }

    func clear() {
        events.removeAll(keepingCapacity: true)
        selectedEventID = nil
        droppedEventCount = 0
    }

    var selectedEvent: CapturedGameEvent? {
        events.first { $0.id == selectedEventID } ?? events.last
    }
}

File: flutter/lib/event_store.dart
import 'package:flutter/foundation.dart';
import 'game_event.dart';

class EventStore extends ChangeNotifier {
  EventStore({this.capacity = 500});
  final int capacity;
  final List<CapturedGameEvent> _events = [];
  int _nextId = 1;
  int? selectedEventId;
  int droppedEventCount = 0;

  List<CapturedGameEvent> get events => List.unmodifiable(_events);
  CapturedGameEvent? get selectedEvent {
    if (_events.isEmpty) return null;
    for (final event in _events) {
      if (event.id == selectedEventId) return event;
    }
    return _events.last;
  }

  void capture(dynamic body, {DateTime? receivedAt}) {
    final event = GameEventParser.parse(
      body,
      id: _nextId++,
      receivedAt: receivedAt,
    );
    _events.add(event);
    if (_events.length > capacity) {
      final overflow = _events.length - capacity;
      _events.removeRange(0, overflow);
      droppedEventCount += overflow;
    }
    selectedEventId = event.id;
    notifyListeners();
  }

  void select(int id) {
    selectedEventId = id;
    notifyListeners();
  }

  void clear() {
    _events.clear();
    selectedEventId = null;
    droppedEventCount = 0;
    notifyListeners();
  }
}
